### PR TITLE
resolve .ts/.tsx/.d.ts first, and then fallback to @types/*

### PR DESCRIPTION
**Pull request by JounQin**

### CODE REVIEW

**Comment by JounQin** [comment]
@alexgorbatchev @bradzacher Do you have any time to review this PR? It is really needed.

**Comment by rdsedmundo** [comment]
+1 for reviewing, really looking forward to having this.

**Comment by JounQin** [comment]
@rdsedmundo You can try `yarn add -D eslint-import-resolver-typescript@JounQin/eslint-import-resolver-typescript#feat/resolve_dts` temporarily.

File: index.js [hunk 1]
```diff
@@ -64,6 +62,24 @@ function resolveFile(source, file, config) {
     foundNodePath = null;
   }
 
+  // naive attempt at @types/* resolution,
+  // iff path is neither absolute nor relative
+  if (
+    /\.jsx?$/.test(foundNodePath) &&
```
[the remark this answers is not in the document]
Comment (JounQin) [reply index.js]: > but after removing this line, my import rules were behaving as expected again.

Can you explain it more specifically? We should only try to resolve `@types/` for `.js/.jsx` files.

[the remark this answers is not in the document]
Comment (trevorblades) [reply index.js]: Yea I installed `@types/mdast` and I want to import types from it using:

```ts
import {Code} from 'mdast';
```

But I'm getting `import/no-unresolved` errors. Do you know an alternative way around this?

[the remark this answers is not in the document]
Comment (trevorblades) [reply index.js]: To add more context, the issue that I'm trying to account for is when `mdast` is unable to be resolved, and `foundNodePath` is set to null in the `try/catch` above this block of code. In those cases, I'd need the import resolver to try to resolve a `@types/*` package.

[the remark this answers is not in the document]
Comment (JounQin) [reply index.js]: So why can not you install `mdast` and `@types/mdast` at the same time? Is that a good practice to only install typings without `.js` source code?

[the remark this answers is not in the document]
Comment (trevorblades) [reply index.js]: The `mdast` package was renamed to `remark` 4 years ago, but `@types/mdast` still contains many types that I need to use. Same goes for `@types/unist`, which contains many useful types but `unist` isn't even an NPM package: https://www.npmjs.com/package/unist

[the remark this answers is not in the document]
Comment (JounQin) [reply index.js]: @trevorblades Maybe we can add an option like `alwaysTryTypes`.

[the remark this answers is not in the document]
Comment (trevorblades) [reply index.js]: Thanks @JounQuin! I'll give it a try. My initial thought is that the condition should look something like this:

```js
(config.alwaysTryTypes && !foundNodePath) || /\.jsx?$/.test(foundNodePath)
```

This way, the code inside the condition will only run when the config option is set _and_ we couldn't resolve the originally-requested package. What do you think?

[the remark this answers is not in the document]
Comment (JounQin) [reply index.js]: Ah, yes, I was just stupid at that moment.

**Comment by trevorblades** [comment]
Works like a charm with the `alwaysFindTypes` option. Thanks @JounQin! 👍

**Comment by Protectator** [comment]
Any news on a potential merge ?

**Comment by JounQin** [comment]
Hey guys, it seems @alexgorbatchev has *'disappeared'* for a few months, or he just forgot this project, so I'm going to fork this repository and republish it as `eslint-import-resolver-ts` to npm. You can simply change your eslint config:

```diff
'import/resolver': {
- typescript
+ ts: {
    alwaysTryTypes: true,
  },
},
```

If anyone is interested, you can follow [eslint-import-resolver-ts](https://github.com/rx-ts/eslint-import-resolver-ts), PR welcome.

And I will not close this PR, in case @alexgorbatchev return.

**Comment by Protectator** [comment]
Great, thanks for taking the initiative.

**Comment by ljharb** [comment]
@JounQin if it's not too late, it's best to publish forks under a scope.

**Comment by trevorblades** [comment]
@ljharb @JounQin you should be able to `unpublish` NPM packages within 48 hours of them being published. Would the scoped name you're referring to be `@jounqin/eslint-import-resolver-typescript`?

**Comment by ljharb** [comment]
Yes, exactly.

**Comment by JounQin** [comment]
@ljharb scoped package shorthand is supported by eslint-plugin-import?

```
{
  '@rxts/typescript': {}
}
```

It's a bit long to me

**Comment by ljharb** [comment]
i don’t believe we support a shorthand, but the priority imo shouldn’t be “what you have to copy and paste into your config file”, but rather “what will make it very clear this is a fork”

**Comment by JounQin** [comment]
@ljharb @trevorblades  I'm not only going to fork it without other changes but also refactor, add new features and try to maintain it in the future.

Of course, if @alexgorbatchev return and accept, I can PR back.

Please help to review https://github.com/rx-ts/eslint-import-resolver-ts/pull/10 if you don't mind.

**Comment by alexgorbatchev** [comment]
Sorry for MIA :( life...

**Comment by JounQin** [comment]
Welcome back @alexgorbatchev . So are you going to maintain this resolver in the future? I can PR back from my fork https://github.com/rx-ts/eslint-import-resolver-ts which has been refactored to be `.ts` based itself which a few new improvements, maybe.

**Comment by alexgorbatchev** [comment]
I think it would make sense to merge back. I can make you a collaborator on the repo. Would that work?

**Comment by JounQin** [comment]
@alexgorbatchev That would be great.

**Comment by alexgorbatchev** [comment]
Done

**Pull request merged**